### ISSUE TITLE
EVP_aes_256_cfb8 NULL pointer dereference if key is empty

### ISSUE THREAD

**Issue by guidovranken** [openssl/openssl#9359]
This is a reproduction of https://oss-fuzz.com/testcase-detail/5732082979962880

Segfault appears to occur in assembly -- this is Intel 64 bit.

```cpp
#include <openssl/evp.h>
#define CF_CHECK_EQ(expr, res) if ( (expr) != (res) ) { goto end; }
#define CF_CHECK_NE(expr, res) if ( (expr) == (res) ) { goto end; }
#define CF_CHECK_GT(expr, res) if ( (expr) <= (res) ) { goto end; }
#define CF_CHECK_GTE(expr, res) if ( (expr) < (res) ) { goto end; }
#define CF_CHECK_LT(expr, res) if ( (expr) >= (res) ) { goto end; }
#define CF_CHECK_LTE(expr, res) if ( (expr) > (res) ) { goto end; }
int main(void)
{
    int len = -1;
    unsigned char iv[16] = {0};
    unsigned char in[16] = {0};
    unsigned char out[1024];
    EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
    const EVP_CIPHER* cipher = EVP_aes_256_cfb8();
    CF_CHECK_EQ(EVP_EncryptInit_ex(ctx, cipher, nullptr, nullptr, nullptr), 1);
    CF_CHECK_EQ(EVP_CIPHER_CTX_set_key_length(ctx, 0), 1);
    CF_CHECK_EQ(EVP_EncryptInit_ex(ctx, nullptr, nullptr, nullptr, iv), 1);
    CF_CHECK_EQ(EVP_EncryptUpdate(ctx, out, &len, in, 16), 1);
end:
    return 0;
}
```

**Comment by mattcaswell** [comment]
Seems to have been caused by #9328.

AES-256-CFB8 does not support variable length keys, and yet after that PR calls to `EVP_CIPHER_CTX_set_key_length` succeed.

Ping @levitte.

**Comment by guidovranken** [comment]
@mattcaswell / others: Just wondering: do you consider Cryptofuzz and its findings beneficial to OpenSSL? Is there anything you would like to see changed or improved?

**Comment by kroeckx** [comment]
It's not always clear in which project the problem is.

**Comment by mattcaswell** [comment]
> @mattcaswell / others: Just wondering: do you consider Cryptofuzz and its findings beneficial to OpenSSL? Is there anything you would like to see changed or improved?

Finding stuff like this is very useful. We're making lots of changes at the moment and it is helpful to find stuff that might otherwise slip through unnoticed. I do agree with @kroeckx though - I tend to ignore the notifications because there's too many of them and its not always clear which project is the cause (or whether it is a problem with cryptofuzz itself). I wait until you raise issues before looking at them.

**Comment by guidovranken** [comment]
Right. If it is a memory bug or undefined behavior, the affected projected can usually be inferred from the stack trace, which you can see on the oss-fuzz.com page for that bug. In case you have doubts whether it concerns OpenSSL master, OpenSSL 1.1.0, OpenSSL 1.0.2, LibreSSL or BoringSSL, you can tell this from the fuzzer name (eg. ```libFuzzer_cryptofuzz_cryptofuzz-openssl-110-noasm```).

If it concerns a different result between two libraries (or two different results from the same library), you can know the involved libraries from the crash name.

Eg. The crash name of https://oss-fuzz.com/testcase-detail/5746833189502976 is ```OpenSSL-OpenSSL-SymmetricDecrypt-CHACHA20_POLYNUMBER-difference```, which means that OpenSSL decrypted into two different plaintexts for CHACHA20-POLY1305 (apparently oss-fuzz changes ```1305``` into ```NUMBER```).

Another example could be ```OpenSSL-libgcrypt-SymmetricEncrypt-AES-difference``` which would mean that OpenSSL and libgcrypt encrypted into a different ciphertext using the AES cipher (with the same plaintext, key, iv).

Furthermore, you can download both the build and the testcase from the oss-fuzz issue page, and run it like this:

```
<binary> --debug <testcase>
```

and it will print some information like parameters (key, iv, ...) and module name (eg. OpenSSL) before running.

But I'll always be stand-by and create Github issues for whatever OSS-Fuzz finds if I have time.

**Comment by paulidale** [comment]
I've tried clicking through to the link in the email notification a couple of times but I don't have permissions.

**Comment by guidovranken** [comment]
@paulidale do you have an email address linked to a Google account? Please give me that and I'll add you.

**Comment by paulidale** [comment]
Email sent, thanks.

**Comment by guidovranken** [comment]
Reminder that this bug is still actual. This new crash: https://oss-fuzz.com/testcase-detail/5696809090940928 probably has the same root cause.

**Comment by mattcaswell** [comment]
Does #9562 solve this?

**Comment by guidovranken** [comment]
> Does #9562 solve this?

Nope -- PoC (see first post in this thread) still crashes.

**Comment by guidovranken** [comment]
Several new crashes at OSS-Fuzz like this one

https://oss-fuzz.com/testcase-detail/5181934937833472

Reports OOB read at 
https://github.com/openssl/openssl/blob/c85d5e025797f4e77c2166150bbf8745e9d00696/crypto/aria/aria.c#L561

Inputs:

```
operation name: SymmetricDecrypt
ciphertext: {0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0xff, 0xff, 0xff, 0xff, 0x20, 0x20, 0x20, 0x20} (16 bytes)
tag: nullopt
aad: nullopt
cipher iv: {0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20} (8 bytes)
cipher key: {0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20} (8 bytes)
cipher: ARIA_192_CCM
cleartextSize: 8224
```

Might be related

**Comment by mattcaswell** [comment]
This should have been fixed by #10449. @guidovranken can you confirm?

**Comment by guidovranken** [comment]
Yes, fixed.
